google/cloud/storage/internal/policy_document_request.cc

```diff
@@ -56,96 +60,154 @@ nlohmann::json TransformConditions(
-  utf32 = conv.from_bytes(utf8_bytes);
-  for (auto c : utf32) {
-    bool is_ascii = EscapeAsciiChar(result, c);
-    if (!is_ascii) {
-      // All unicode characters should be encoded as \udead.
-      std::ostringstream os;
-      os << "\\u" << std::setw(4) << std::setfill('0') << std::hex
-         << static_cast<std::uint32_t>(c);
-      result.append(os.str());
-    }
-  }
-  return result;
+StatusOr<std::string> Escape3(absl::string_view s, std::size_t pos) {
+  auto status = ValidateUTF8Encoding(s, pos, 3);
+  if (!status.ok()) return status;
+  auto const e = s.substr(pos, 3);
+  return absl::StrFormat(
+      "\\u%04x", DecodeUTF8(0, Header3(e[0]), Trailer(e[1]), Trailer(e[2])));
 }
-#else   // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
-StatusOr<std::string> PostPolicyV4EscapeUTF8(std::string const&) {
-  return Status(StatusCode::kUnimplemented,
-                "Signing POST policies is unavailable with this compiler due "
-                "to the lack of exception support.");
+
+StatusOr<std::string> Escape4(absl::string_view s, std::size_t pos) {
+  auto status = ValidateUTF8Encoding(s, pos, 4);
+  if (!status.ok()) return status;
+  auto const e = s.substr(pos, 4);
+  auto codepoint =
+      DecodeUTF8(Header4(e[0]), Trailer(e[1]), Trailer(e[2]), Trailer(e[3]));
+  if (codepoint <= 0xFFFFU) return absl::StrFormat("\\u%04x", codepoint);
+  return absl::StrFormat("\\U%08x", codepoint);
 }
-#endif  // GOOGLE_CLOUD_CPP_HAVE_EXCEPTIONS
 
-StatusOr<std::string> PostPolicyV4Escape(std::string const& utf8_bytes) {
+StatusOr<std::string> EscapeUTF8(absl::string_view s) {
+  using Encoder =
+      absl::FunctionRef<StatusOr<std::string>(absl::string_view, std::size_t)>;
+  struct {
+    std::uint32_t mask;
+    std::uint32_t value;
+    Encoder encode;
+  } const encodings[] = {
+      {kMask1, 0b0000'0000U, Escape1},
+      {kMask2, 0b1100'0000U, Escape2},
+      {kMask3, 0b1110'0000U, Escape3},
+      {kMask4, 0b1111'0000U, Escape4},
+  };
+  // Iterate over all the bytes in the input string. Interpreting each UTF-8
+  // sequence as needed.
   std::string result;
-
-  for (char32_t c : utf8_bytes) {
-    bool is_ascii = EscapeAsciiChar(result, c);
-    if (!is_ascii) {
-      // We first try to escape the string assuming it's plain ASCII. If it
-      // turns out that there are non-ASCII characters, we fall back to
-      // interpreting it as proper UTF-8. We do it because it will be faster in
-      // the common case and, more importantly, this allows the common case to
-      // work properly on compilers which don't have UTF8 support.
-      return PostPolicyV4EscapeUTF8(utf8_bytes);
+  std::size_t pos = 0;
+  while (pos != s.size()) {
+    // Test if s[pos] is a 1, 2, 3 or 4 byte UTF-8 codepoint. If it is matched
+    // add the escaped characters to `result`. If it is not matched return an
+    // error.
+    bool matched = false;
+    std::size_t n = 1;
+    for (auto const& e : encodings) {
+      if ((e.mask & s[pos]) == e.value) {
```

Comment:
The four values are direct functions of their respective masks, so ...
```
      if ((s[pos] & e.mask) == ((e.mask - 1) & e.mask) {
```
and get rid of `e.value`.

Reply:
Refactored to a function.